### ISSUE TITLE
http3-server crashes

### ISSUE THREAD

**Issue by bagder** [cloudflare/quiche#1119]
The http3-server example (built from git at commit af678919d3) code crashes whenever I try to access it from curl. Using curl with quiche or ngtcp2 doesn't seem to make a difference. Here's a crash when I use curl+ngtcp2:

## server side
~~~
$ ./http3-server localhost 8443
stateless retry
sent 113 bytes
recv would block
copy 28 bytes
=================================================================
==607859==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff99b2ec68 at pc 0x7f9a874de9b0 bp 0x7fff99b2eba0 sp 0x7fff99b2e350
READ of size 28 at 0x7fff99b2ec68 thread T0
    #0 0x7f9a874de9af in __interceptor_memcpy ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:827
    #1 0x563f1229c450 in create_conn /home/daniel/src/quiche/examples/http3-server.c:209
    #2 0x563f122a06ed in recv_cb /home/daniel/src/quiche/examples/http3-server.c:350
    #3 0x7f9a87498632 in ev_invoke_pending (/usr/lib/x86_64-linux-gnu/libev.so.4+0x5632)
    #4 0x7f9a8749be70 in ev_run (/usr/lib/x86_64-linux-gnu/libev.so.4+0x8e70)
    #5 0x563f1229b974 in ev_loop /usr/include/ev.h:841
    #6 0x563f122a3743 in main /home/daniel/src/quiche/examples/http3-server.c:587
    #7 0x7f9a8716a7ec in __libc_start_main ../csu/libc-start.c:332
    #8 0x563f1229b899 in _start (/home/daniel/src/quiche/examples/http3-server+0x5b899)

Address 0x7fff99b2ec68 is located in stack of thread T0 at offset 40 in frame
    #0 0x563f1229c032 in create_conn /home/daniel/src/quiche/examples/http3-server.c:181

  This frame has 1 object(s):
    [32, 40) 'peer_addr' (line 180) <== Memory access at offset 40 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:827 in __interceptor_memcpy
Shadow bytes around the buggy address:
  0x10007335dd30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007335dd40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007335dd50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007335dd60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007335dd70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10007335dd80: 00 00 00 00 00 00 00 00 f1 f1 f1 f1 00[f3]f3 f3
  0x10007335dd90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10007335dda0: 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1 01 f2
  0x10007335ddb0: 04 f2 04 00 00 f2 00 00 00 f2 00 00 00 f2 00 00
  0x10007335ddc0: 00 f2 00 00 00 f2 f2 f2 00 00 f2 f2 00 00 f2 f2
  0x10007335ddd0: 00 00 04 f2 f2 f2 f2 f2 00 00 04 f2 f2 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==607859==ABORTING
~~~

## client side
~~~
$ ./src/curl --http3 https://localhost:8443/ -v
* STATE: INIT => CONNECT handle 0x55771fefdba8; line 1835 (connection #-5000)
* Added connection 0. The cache now contains 1 members
* family0 == v4, family1 == v6
*   Trying 127.0.0.1:8443...
* Connect socket 5 over QUIC to 127.0.0.1:8443
* STATE: CONNECT => CONNECTING handle 0x55771fefdba8; line 1896 (connection #0)
* ngtcp2: recvfrom() unexpectedly returned -1
* connect to 127.0.0.1 port 8443 failed: Connection refused
*   Trying ::1:8443...
* Connect socket 6 over QUIC to ::1:8443
* ngtcp2: recvfrom() unexpectedly returned -1
* connect to ::1 port 8443 failed: Connection refused
* Failed to connect to localhost port 8443 after 1007 ms: Connection refused
* multi_done: status: 7 prem: 1 done: 0
* The cache now contains 0 members
* Closing connection 0
* Expire cleared (transfer 0x55771fefdba8)
curl: (7) ngtcp2: recvfrom() unexpectedly returned -1
~~~

**Comment by LPardue** [comment]
Weird. I've just tried with a clean build of curl and http3-server and don't see this problem. 

The project layout changed recently and I wonder if there is some hangover issue. It seems like you're running the binary from `/home/daniel/src/quiche/examples/http3-server` but for me I had to run it from the equivalent of `/home/lucas/stuff/quiche/quiche/examples/http3-server`

What's also strange is the line `copy 28 bytes`. I don't see anywhere obvious where that would be generated in the HEAD commit.
